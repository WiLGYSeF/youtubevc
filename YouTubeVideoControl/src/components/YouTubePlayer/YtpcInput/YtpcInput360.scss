.three-sixty {
  input[type="number"] {
    width: 40px;
  }
}
