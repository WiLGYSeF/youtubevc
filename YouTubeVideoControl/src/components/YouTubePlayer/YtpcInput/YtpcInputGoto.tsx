import React, { useEffect } from 'react';

import TimestampInput from 'components/common/TimestampInput';
import { YtpcGotoState } from 'objects/YtpcEntry/YtpcGotoEntry';
import secondsToTimestamp from 'utils/secondsToTimestamp';
import useStatePropBacked from 'utils/useStatePropBacked';
import { YtpcControlInput } from './YtpcControlInput';

function YtpcInputGoto(props: YtpcControlInput) {
  const pstate = props.state as YtpcGotoState;
  const [gotoTime, setGotoTime] = useStatePropBacked(pstate?.gotoTime ?? 0);

  useEffect(() => {
    const state: YtpcGotoState = {
      gotoTime,
    };
    props.setControlInputState(state);
  }, [gotoTime]);

  return (
    <TimestampInput
      value={secondsToTimestamp(gotoTime)}
      onChange={setGotoTime}
    />
  );
}

export default YtpcInputGoto;
