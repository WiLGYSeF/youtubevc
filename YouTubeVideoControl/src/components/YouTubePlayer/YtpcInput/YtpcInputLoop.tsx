import React, { useEffect } from 'react';

import Checkbox from 'components/common/Checkbox/Checkbox';
import NumberInput from 'components/common/NumberInput/NumberInput';
import TimestampInput from 'components/common/TimestampInput/TimestampInput';
import { YtpcLoopState } from 'objects/YtpcEntry/YtpcLoopEntry';
import useStatePropBacked from 'utils/useStatePropBacked';
import { YtpcControlInput } from './YtpcControlInput';

import './YtpcInputLoop.scss';

const LOOP_COUNT_DEFAULT = 1;

function YtpcInputLoop(props: YtpcControlInput) {
  const pstate = props.state as YtpcLoopState;
  const [loopBackTo, setLoopBackTo] = useStatePropBacked(pstate?.loopBackTo ?? 0);
  const [forever, setForever] = useStatePropBacked((pstate?.loopCount ?? -1) < 0);
  const [loopCount, setLoopCount] = useStatePropBacked(pstate?.loopCount ?? LOOP_COUNT_DEFAULT);

  useEffect(() => {
    const state: YtpcLoopState = {
      atTime: pstate.atTime,
      controlType: pstate.controlType,
      loopBackTo,
      loopCount: forever ? -1 : loopCount,
    };
    props.setEntryState(state);
  }, [loopBackTo, forever, loopCount]);

  return (
    <div className="loop">
      <TimestampInput
        value={loopBackTo}
        onChange={setLoopBackTo}
      />
      <Checkbox
        label={forever ? 'forever' : 'times'}
        checked={forever}
        onChange={(checked: boolean) => {
          if (!checked) {
            setLoopCount(LOOP_COUNT_DEFAULT);
          }
          setForever(checked);
        }}
      />
      {!forever && (
        <span className="loop-count">
          <NumberInput
            minValue={0}
            value={loopCount}
            forceValue
            onChange={setLoopCount}
          />
        </span>
      )}
    </div>
  );
}

export default YtpcInputLoop;
