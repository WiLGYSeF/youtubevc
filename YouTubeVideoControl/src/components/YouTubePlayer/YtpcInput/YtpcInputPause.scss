.pause {
  input {
    width: 45px;
  }
}
