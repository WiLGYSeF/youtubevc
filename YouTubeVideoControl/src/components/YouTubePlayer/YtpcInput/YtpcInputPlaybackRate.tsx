import React, { ChangeEvent, useEffect } from 'react';

import { ControlType } from 'objects/YtpcEntry/YouTubePlayerControllerEntry';
import { YtpcPlaybackRateState } from 'objects/YtpcEntry/YtpcPlaybackRateEntry';
import useStatePropBacked from 'utils/useStatePropBacked';
import { YtpcControlInput } from './YtpcControlInput';

interface YtpcInputPlaybackRateProps extends YtpcControlInput {
  playbackRates?: number[];
}

const PLAYBACK_RATE_DEFAULT = 1;

function YtpcInputPlaybackRate(props: YtpcInputPlaybackRateProps) {
  const pstate = props.defaultState as YtpcPlaybackRateState;
  const dPlaybackRate = pstate.playbackRate ?? PLAYBACK_RATE_DEFAULT;

  const [playbackRate, setPlaybackRate] = useStatePropBacked(dPlaybackRate);
  const playbackRates = props.playbackRates ?? [PLAYBACK_RATE_DEFAULT];

  useEffect(() => {
    const state: YtpcPlaybackRateState = {
      atTime: props.entryState.atTime,
      controlType: ControlType.PlaybackRate,
      playbackRate,
    };
    props.setEntryState(state);
  }, [playbackRate]);

  return (
    <div data-testid="playback-rate">
      <select
        value={playbackRate}
        onChange={(e: ChangeEvent<HTMLSelectElement>) => {
          setPlaybackRate(Number(e.target.value));
        }}
      >
        {playbackRates.map((r) => <option key={r} value={r}>{r}</option>)}
      </select>
    </div>
  );
}

export default YtpcInputPlaybackRate;
