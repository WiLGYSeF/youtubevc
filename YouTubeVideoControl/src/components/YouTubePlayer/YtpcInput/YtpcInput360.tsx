import React, { useEffect } from 'react';

import Checkbox from 'components/common/Checkbox/Checkbox';
import NumberInput from 'components/common/NumberInput/NumberInput';
import Ytpc360Entry, { Ytpc360State } from 'objects/YtpcEntry/Ytpc360Entry';
import { ControlType } from 'objects/YtpcEntry/YouTubePlayerControllerEntry';
import useStatePropBacked from 'utils/useStatePropBacked';
import { YtpcControlInput } from './YtpcControlInput';

import './YtpcInput360.scss';

export const LERP_TIME_DEFAULT = 3;

function YtpcInput360(props: YtpcControlInput) {
  const pstate = props.defaultState as Ytpc360State;
  const dYaw = pstate?.sphereProps?.yaw ?? Ytpc360Entry.YAW_DEFAULT;
  const dPitch = pstate?.sphereProps?.pitch ?? Ytpc360Entry.PITCH_DEFAULT;
  const dRoll = pstate?.sphereProps?.roll ?? Ytpc360Entry.ROLL_DEFAULT;
  const dFov = pstate?.sphereProps?.fov ?? Ytpc360Entry.FOV_DEFAULT;
  const dLerpSeconds = pstate?.lerpSeconds ?? -1;

  const [yaw, setYaw] = useStatePropBacked(dYaw);
  const [pitch, setPitch] = useStatePropBacked(dPitch);
  const [roll, setRoll] = useStatePropBacked(dRoll);
  const [fov, setFov] = useStatePropBacked(dFov);

  const [lerpSet, setLerp] = useStatePropBacked((pstate?.lerpSeconds ?? -1) >= 0);
  const [lerpSeconds, setLerpSeconds] = useStatePropBacked(dLerpSeconds >= 0
    ? dLerpSeconds
    : LERP_TIME_DEFAULT);

  useEffect(() => {
    const state: Ytpc360State = {
      atTime: pstate.atTime,
      controlType: ControlType.ThreeSixty,
      sphereProps: {
        yaw: yaw === 360 ? 0 : yaw,
        pitch,
        roll,
        fov,
      },
      lerpSeconds: lerpSet ? lerpSeconds : -1,
    };
    props.setEntryState(state);
  }, [yaw, pitch, roll, fov, lerpSet, lerpSeconds]);

  return (
    <div className="three-sixty">
      <span className="yaw">
        <NumberInput
          label="yaw: "
          minValue={Ytpc360Entry.YAW_MIN} maxValue={Ytpc360Entry.YAW_MAX} step={null}
          defaultValue={dYaw}
          clamp={false}
          forceValue
          onChange={setYaw}
        />
      </span>
      <span className="pitch">
        <NumberInput
          label="pitch: "
          minValue={Ytpc360Entry.PITCH_MIN} maxValue={Ytpc360Entry.PITCH_MAX} step={null}
          defaultValue={dPitch}
          clamp={false}
          forceValue
          onChange={setPitch}
        />
      </span>
      <span className="roll">
        <NumberInput
          label="roll: "
          minValue={Ytpc360Entry.ROLL_MIN} maxValue={Ytpc360Entry.ROLL_MAX} step={null}
          defaultValue={dRoll}
          clamp={false}
          forceValue
          onChange={setRoll}
        />
      </span>
      <span className="fov">
        <NumberInput
          label="fov: "
          minValue={Ytpc360Entry.FOV_MIN} maxValue={Ytpc360Entry.FOV_MAX} step={null}
          defaultValue={dFov}
          forceValue
          onChange={setFov}
        />
      </span>
      <span className="lerp">
        <Checkbox
          label="lerp"
          defaultChecked={lerpSet}
          onChange={setLerp}
        />
      </span>
      <span
        className="lerp-seconds" style={{
          display: lerpSet ? '' : 'none',
        }}
      >
        <span> for </span>
        <NumberInput
          label=" seconds"
          labelRight
          minValue={0} step={null}
          defaultValue={dLerpSeconds >= 0 ? dLerpSeconds : LERP_TIME_DEFAULT}
          forceValue
          onChange={setLerpSeconds}
        />
      </span>
    </div>
  );
}

export default YtpcInput360;
