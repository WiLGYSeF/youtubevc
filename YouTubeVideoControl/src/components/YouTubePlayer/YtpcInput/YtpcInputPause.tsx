import React, { useEffect } from 'react';

import TimestampInput from 'components/common/TimestampInput';
import { YtpcPauseState } from 'objects/YtpcEntry/YtpcPauseEntry';
import useStatePropBacked from 'utils/useStatePropBacked';
import { YtpcControlInput } from './YtpcControlInput';

import './YtpcInputPause.scss';

const PAUSE_FOR_DEFAULT = 5;

function YtpcInputPause(props: YtpcControlInput) {
  const pstate = props.state as YtpcPauseState;
  const [pauseFor, setPauseFor] = useStatePropBacked(pstate?.pauseTime ?? PAUSE_FOR_DEFAULT);

  useEffect(() => {
    const state: YtpcPauseState = {
      atTime: pstate.atTime,
      controlType: pstate.controlType,
      pauseTime: pauseFor,
    };
    props.setEntryState(state);
  }, [pauseFor]);

  return (
    <div className="pause">
      <TimestampInput
        value={pstate.pauseTime}
        onChange={setPauseFor}
      />
    </div>
  );
}

export default YtpcInputPause;
