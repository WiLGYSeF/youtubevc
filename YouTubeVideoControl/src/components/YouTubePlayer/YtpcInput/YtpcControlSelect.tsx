import React, { ChangeEvent, useEffect } from 'react';

import { ControlType } from 'objects/YtpcEntry/YouTubePlayerControllerEntry';
import Ytpc360Entry from 'objects/YtpcEntry/Ytpc360Entry';
import YtpcGotoEntry from 'objects/YtpcEntry/YtpcGotoEntry';
import YtpcLoopEntry from 'objects/YtpcEntry/YtpcLoopEntry';
import YtpcPauseEntry from 'objects/YtpcEntry/YtpcPauseEntry';
import YtpcPlaybackRateEntry from 'objects/YtpcEntry/YtpcPlaybackRateEntry';
import YtpcVolumeEntry from 'objects/YtpcEntry/YtpcVolumeEntry';
import YtpcInputVolume from './YtpcInputVolume';
import YtpcInputPlaybackRate from './YtpcInputPlaybackRate';
import YtpcInputPause from './YtpcInputPause';
import YtpcInputLoop from './YtpcInputLoop';
import YtpcInputGoto from './YtpcInputGoto';
import YtpcInput360 from './YtpcInput360';

interface ControlSelectProps {
  is360Video: boolean;
  controlInputType: ControlType;
  setControlInput(type: ControlType, component: (props: any) => JSX.Element): void;
}

type Control = {
  enabled: boolean;
  text: string;
  component: (props: any) => JSX.Element;
};

function YtpcControlSelect(props: ControlSelectProps) {
  const controls = new Map<string, Control>([
    [ControlType.ThreeSixty, {
      // enabled: props.is360Video, // only set after video start playing, keep enabled for now
      enabled: true,
      text: Ytpc360Entry.ACTION_STR,
      component: YtpcInput360,
    }],
    [ControlType.Goto, {
      enabled: true,
      text: YtpcGotoEntry.ACTION_STR,
      component: YtpcInputGoto,
    }],
    [ControlType.Loop, {
      enabled: true,
      text: YtpcLoopEntry.ACTION_STR,
      component: YtpcInputLoop,
    }],
    [ControlType.Pause, {
      enabled: true,
      text: YtpcPauseEntry.ACTION_STR,
      component: YtpcInputPause,
    }],
    [ControlType.PlaybackRate, {
      enabled: true,
      text: YtpcPlaybackRateEntry.ACTION_STR,
      component: YtpcInputPlaybackRate,
    }],
    [ControlType.Volume, {
      enabled: true,
      text: YtpcVolumeEntry.ACTION_STR,
      component: YtpcInputVolume,
    }],
  ]);

  const setControlInput = (type: ControlType) => {
    const control = controls.get(type) as Control;
    props.setControlInput(type, control.component);
  };

  useEffect(() => {
    setControlInput(props.controlInputType);
  }, [props.controlInputType]);

  return (
    <select
      value={props.controlInputType}
      onChange={(e: ChangeEvent<HTMLSelectElement>) => {
        setControlInput(e.target.value as ControlType);
      }}
    >
      {Array.from(controls.entries()).map((c) => (
        <option key={c[0]} value={c[0]} disabled={!c[1].enabled}>
          {c[1].text}
        </option>
      ))}
    </select>
  );
}

export default YtpcControlSelect;
