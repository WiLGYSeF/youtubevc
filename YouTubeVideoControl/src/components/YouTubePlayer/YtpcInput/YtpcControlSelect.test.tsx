import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ControlType } from 'objects/YtpcEntry/YouTubePlayerControllerEntry';
import YtpcControlSelect from './YtpcControlSelect';
import YtpcInput360 from './YtpcInput360';
import YtpcInputGoto from './YtpcInputGoto';
import YtpcInputLoop from './YtpcInputLoop';
import YtpcInputPause from './YtpcInputPause';
import YtpcInputPlaybackRate from './YtpcInputPlaybackRate';
import YtpcInputVolume from './YtpcInputVolume';

function getInputs(container: HTMLElement): ({
  select: HTMLSelectElement,
}) {
  return {
    select: container.getElementsByTagName('select')[0],
  };
}

describe('YtpcControlSelect', () => {
  it('shows all input options', () => {
    const { container } = render(<YtpcControlSelect
      is360Video={false}
      defaultControlType={ControlType.Goto}
      setControlInput={() => {}}
    />);

    const { select } = getInputs(container);
    const options = Array.from(select.getElementsByTagName('option'));

    expect(options.map((opt) => opt.value)).toEqual([
      ControlType.ThreeSixty,
      ControlType.Goto,
      ControlType.Loop,
      ControlType.Pause,
      ControlType.PlaybackRate,
      ControlType.Volume,
    ]);
  });

  it.each([
    [ControlType.ThreeSixty, YtpcInput360],
    [ControlType.Goto, YtpcInputGoto],
    [ControlType.Loop, YtpcInputLoop],
    [ControlType.Pause, YtpcInputPause],
    [ControlType.PlaybackRate, YtpcInputPlaybackRate],
    [ControlType.Volume, YtpcInputVolume],
  ])(
    'returns the %s input component',
    (controlType: ControlType, expected: Function) => {
      const setControlInputMock = jest.fn();

      const { container } = render(<YtpcControlSelect
        is360Video={false}
        defaultControlType={ControlType.Goto}
        setControlInput={setControlInputMock}
      />);

      const { select } = getInputs(container);

      setControlInputMock.mockClear();
      userEvent.selectOptions(select, controlType);

      expect(setControlInputMock.mock.calls[0][0]).toEqual(controlType);
      expect(setControlInputMock.mock.calls[0][1]).toEqual(expected);

      expect(select.value).toEqual(controlType);
    },
  );

  it('updates value on props change', () => {
    let controlType = ControlType.Goto;
    const setControlInputMock = jest.fn();

    const { container, rerender } = render(<YtpcControlSelect
      is360Video={false}
      defaultControlType={controlType}
      setControlInput={setControlInputMock}
    />);

    const { select } = getInputs(container);

    expect(select.value).toEqual(controlType);

    controlType = ControlType.Loop;

    rerender(<YtpcControlSelect
      is360Video={false}
      defaultControlType={controlType}
      setControlInput={setControlInputMock}
    />);

    expect(select.value).toEqual(controlType);
  });
});
