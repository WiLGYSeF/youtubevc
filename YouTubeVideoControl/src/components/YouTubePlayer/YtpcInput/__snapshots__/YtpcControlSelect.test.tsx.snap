// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YtpcControlSelect shows all input options 1`] = `
<select
  onChange={[Function]}
  value="goto"
>
  <option
    disabled={false}
    value="360"
  >
    set 360Â° view to
  </option>
  <option
    disabled={false}
    value="goto"
  >
    go to
  </option>
  <option
    disabled={false}
    value="loop"
  >
    loop back to
  </option>
  <option
    disabled={false}
    value="pause"
  >
    pause for
  </option>
  <option
    disabled={false}
    value="playback-rate"
  >
    set playback rate to
  </option>
  <option
    disabled={false}
    value="volume"
  >
    set volume to
  </option>
</select>
`;
